### PR TITLE
Tensor preprocessing & Lora fix

### PR DESCRIPTION
**Pull request by 1larity**
## PR: Improve Training Reliability on Low-VRAM Systems

Resolves multiple OOM, device residency, feature extraction issues.
#292 #308 #167 
---

This PR improves **training reliability on low-VRAM systems** by:

- Automatically switching **quantized inference presets** to a **training-safe, non-quantized configuration**

- Hardening **CPU/GPU tensor residency and dtype handling** to prevent device mismatch crashes during preprocessing and LoRA training

-Feature added, when training debug flag is set DEBUG_TRAINING = "VERBOSE", Epoch may be set below 100 for quick test cycles.

---

### **LoRA training could run against quantized DiT and fail**
- **Bug:** LoRA injection (PEFT) is incompatible with quantized decoder modules in this runtime.
- **Fix:** Added quantization guardrails and a best-effort auto-switch to a **training preset with quantization disabled**, with clear user-facing status and error messages.

---

### **Training flow could proceed with invalid training state input**
- **Bug:** Training state could be non-dict and break downstream assumptions.
- **Fix:** Added a **state normalization fallback** in the training event wrapper.

---

### **TensorBoard logger initialization could hard-fail training startup**
- **Bug:** Missing TensorBoard modules could cause logger setup to abort training.
- **Fix:** Logger creation is now **optional**; training proceeds without TensorBoard if unavailable.

---

### **Preprocessing encoder path could crash from CPU/CUDA mismatch**
- **Bug:** Encoder inputs could reside on a different device or dtype than the active DiT model, causing backend or flash-attention failures.
- **Fix:** Run the encoder under the **model load context** and explicitly align all relevant tensors to the model’s device and dtype before execution.

---

### **Lyrics encoding could fail due to device mismatch**
- **Bug:** Token tensors could be created on a device different from the text encoder parameters, causing `embedding` / `index_select` failures.
- **Fix:** Explicitly align lyric input tensors to the **actual text encoder device**.

---

### **Training UI could report success after internal failure**
- **Bug:** Certain failure states were treated as normal progress or completion.
- **Fix:** Added **failure detection** in the progress loop and an explicit **failed finalization path**.

---

### **VAE dtype/device handling unstable for low-VRAM offload paths**
- **Bug:** CPU-offloaded VAE dtype handling was not tier-aware and could be slow or fragile.
- **Fix:** Added **tier-aware CPU dtype selection** (e.g. `float32` on lower-VRAM tiers) and explicit VAE offload dtype handling.

---

### **No reusable initialization context for preset switching**
- **Bug:** Training-safe reinitialization could not reliably occur without access to prior init parameters.
- **Fix:** Persist the **last successful initialization parameters** and reuse them in `switch_to_training_preset()`.

---

## Key Files Updated

- `trainer.py`
- `preprocess.py`
- `preprocess_lyrics.py`
- `training_handlers.py`
- `training.py`
- `handler.py`
- `__init__.py`

---

## User Impact

- **More robust training startup** on constrained GPUs
- **Fewer device/dtype mismatch crashes** during preprocessing and training
- **Clearer and earlier error reporting** for unsupported training configurations
- **Graceful behavior** when optional logging dependencies are missing

---

Copilot review: response and rebuttals

_Pre-check note_: pre-existing mojibake / encoding artifacts are **out of scope** for this commit review and should be ignored. That said, **symbol-based dynamic string matching is fragile** and should be minimized where possible.

---

## **High Severity**

_Copilot comment_: **trainer.py (lines 230–237)** — LoRA fail-fast with quantization enabled should auto-recover.  
_Response_: **Rejected (not a bug for this commit)**. The UI path already attempts automatic recovery via `switch_to_training_preset()`. Trainer-level fail-fast behavior is **intentional defensive behavior** for non-UI callers.

---

_Copilot comment_: **preprocess.py (lines 127–160)** — `next(model.parameters())` may fail if the model has no parameters.  
_Response_: **Accepted (hardening follow-up)**. Not an observed regression, but we should add a **guarded fallback** for `StopIteration` / uninitialized-model edge cases.

---

_Copilot comment_: **handler.py (lines 534–566)** — `last_init_params` reliance could cause silent failure.  
_Response_: **Rejected**. Failure is **explicit** (`False` return + status message), and `last_init_params` is persisted only after **successful initialization by design**.

---

## **Medium Severity**

_Copilot comment_: **training_handlers.py (lines 549–694)** — Low-VRAM warnings are not actionable enough.  
_Response_: **Accepted (UX improvement, non-blocking)**. We should append **concrete mitigation tips** (e.g. smaller batch size, gradient accumulation, disabling optional components).

---

_Copilot comment_: **preprocess_lyrics.py (lines 13–20)** — Assumes single-device text encoder.  
_Response_: **Pre-existing architectural limitation** (single-device / offload lifecycle). Not a regression introduced by this patch.

---

_Copilot comment_: **handler.py (lines 738–780)** — Missing `model_name` validation may cause silent failures.  
_Response_: **Pre-existing behavior, low risk in this context**. Call sites use fixed internal names; this commit only adjusted **VAE dtype handling on offload**.

---

## **Low Severity**

_Copilot comment_: **training.py (lines 24–427)** — Debug epoch defaults may be inconsistent if toggled mid-session.  
_Response_: **Rejected (not a regression)**. This is **config-at-load behavior**, not a new runtime toggle path introduced by this commit.

---

_Copilot comment_: **trainer.py (lines 285–296)** — TensorBoard warning should suggest install command.  
_Response_: **Accepted (minor polish)**. Add `pip install tensorboard` guidance to the warning text.

---

## **Regression Checks Summary**

_Copilot comment_: Reliance on `last_init_params`, user-visible warnings, and multi-device assumptions need follow-up.  
_Response_: **Partially accepted**.  
- Actionable warning text and preprocess parameter-guard hardening are valid follow-ups.  
- `last_init_params` silent-failure concern is **not valid**.  
- Multi-device concerns are **pre-existing and out of scope** for this patch.

---

### CODE REVIEW

**Comment by 1larity** [comment]
@ChuxiJ Mojibake issue should really be addressed, multiple places in existing code are fragile due to use of Mojibake in string compares.

Repo-wide scan found these fragile status-string comparisons:

training_handlers.py (line 646)
status_text.startswith("âŒ") (mojibake literal compare)
training_handlers.py (line 647)
status_text.startswith("❌")
training_handlers.py (line 648)
"training failed" in status_lower
training_handlers.py (line 649)
"error:" in status_lower
training_handlers.py (line 650)
"module not found" in status_lower
api_server.py (line 1464)
audio_codes.startswith("❌")
llm_inference.py (line 413)
status_msg.startswith("❌")
label_utils.py (line 15)
not codes_string.startswith("❌")
results_handlers.py (line 1110)
not alignment_report.startswith("\n⚠️")
results_handlers.py (line 1137)
not alignment_report.startswith("\n⚠️")

Also relevant (not compare, but encoding-damaged text): many mojibake UI strings in training_handlers.py (for example around lines 45, 353-356, 540+, 703, 715).

locale/encoding can impact this:

If text is decoded with wrong codepage (for example cp1252 vs UTF-8), emoji markers become mojibake and compares fail.
Unicode variation selectors (for example ⚠ vs ⚠️) can fail startswith even when visually similar.

Locale itself is less the core issue than I/O encoding + normalization + codepage at boundaries.

File: acestep/gradio_ui/events/training_handlers.py [hunk 1]
```diff
@@ -551,6 +550,37 @@ def start_training(
...
+                "",
+                None,
+                training_state,
+            )
+
+        yield "Switching model to training preset (disable quantization)...", "", None, training_state
+        if hasattr(dit_handler, "switch_to_training_preset"):
+            switch_status, switched = dit_handler.switch_to_training_preset()
+            if not switched:
+                yield f"ï¿½ {switch_status}", "", None, training_state
```
Comment (ChuxiJ) [inline acestep/gradio_ui/events/training_handlers.py:577]: There seems to be an issue with the demo on Gradio, as some people have reported it. I made changes halfway through. Are you sure you want to continue using this character?

Comment (ChuxiJ) [reply acestep/gradio_ui/events/training_handlers.py:577]: use❌ or ⚠️ or ✅

**Review by ChuxiJ — APPROVED** [review]
